controller/api/destination/external-workload/endpoints_reconciler.go

```diff
@@ -0,0 +1,624 @@
+	//
+	// We do not segment slices based on ports since we do not support a service
+	// with more than 100 ports.
+	toReconcile := []*discoveryv1.EndpointSlice{}
+	for _, slice := range epSlices {
+		// Loop through the endpointslices and figure out which endpointslice
+		// does not have an ownerRef set to the service. If a slice has been
+		// selected but does not point to the service, we delete it.
+		if !ownedBy(slice, svc) {
+			toDelete = append(toDelete, slice)
+		} else {
+			toReconcile = append(toReconcile, slice)
+		}
+	}
+
+	// Build a list of endpoints we want to create / update. This will be based
+	// off the external workloads we have read.
+	// We use an EndpointSet from the upstream util library. Each Endpoint we
+	// add will be hashed internally.
+	desiredEndpoints := epsliceutil.EndpointSet{}
+
+	// A PortMapKey is an upstream type that when created creates a hash of a
+	// port list. We use a map to ensure we don't add ports twice (i.e. we add
+	// them to a set).
+	desiredEndpointPortsSet := map[epsliceutil.PortMapKey][]discoveryv1.EndpointPort{}
+	for _, extWorkload := range extWorkloads {
+		// We skip workloads with no IPs
+		if len(extWorkload.Spec.WorkloadIPs) == 0 {
+			continue
+		}
+
+		// Find which ports a service selects (or maps to) on an external workload
+		ports, err := r.findEndpointPorts(svc, extWorkload)
+		// Since we require a workload to document its ports, if a service's
+		// targetPort does not match any ports on a workload, we skip it.
+		//
+		// This is different to the upstream where a container port needn't be
+		// documented
+		if err != nil {
+			r.log.Debugf("skipping workload; failed to remap port for service %s/%s and workload %s/%s: %v", svc.Namespace, svc.Name, extWorkload.Namespace, extWorkload.Name, err)
+			continue
+		}
+
+		portHash := epsliceutil.NewPortMapKey(ports)
+		if _, ok := desiredEndpointPortsSet[portHash]; !ok {
+			desiredEndpointPortsSet[portHash] = ports
+		}
+
+		ep := externalWorkloadToEndpoint(discoveryv1.AddressTypeIPv4, extWorkload, svc)
+		desiredEndpoints.Insert(&ep)
+	}
+
+	desiredEndpointPorts := []discoveryv1.EndpointPort{}
+	for _, ports := range desiredEndpointPortsSet {
+		desiredEndpointPorts = append(desiredEndpointPorts, ports...)
+	}
+
+	// If there are any slices whose ports no longer match what we want in our
+	// current reconciliation, delete them
+	//
+	// Note: in the upstream they run some more complicated diffing before
+	// applying to ensure creates & deletes turn into updates. We simplify and
+	// instead choose to re-create the slice if ports have changed.
+	for _, currentSlice := range epSlices {
+		portHash := epsliceutil.NewPortMapKey(currentSlice.Ports)
+		if _, ok := desiredEndpointPortsSet[portHash]; !ok {
+			toDelete = append(toDelete, currentSlice)
+		}
+	}
+
+	if len(desiredEndpointPorts) > maxEndpointPortsQuota {
+		// TODO: we should ensure this is bubbled up and visible to users, not
+		// sure the error msg makes sense.
```

Comment:
TODO: may want a counter for this too

Reply:
Port quota's no longer a thing since we now segment by ports.